File: Questão-5/Animal.go
package main

import (
  "fmt"
)


type AnimalInter interface {
  EmitirSom() string
}


type Animal struct {
  Nome    string
  Especie string
}

type Cachorro struct {
  Animal
}

func (c Cachorro) EmitirSom() string {
  return "AuAu"
}

type Gato struct {
  Animal
}

func (g Gato) EmitirSom() string {
  return "Miau"
}

type Vaca struct{
  Animal
}

func (v Vaca) EmitirSom() string {
  return "Muuh"
}

type GrupoDeAnimais struct {
  animais []AnimalInter
}


func (g *GrupoDeAnimais) AdicionarAnimal(animal AnimalInter) {
  g.animais = append(g.animais, animal)
}


func (g GrupoDeAnimais) EmitirSons() {
  for _, animal := range g.animais {
    fmt.Println("Som:", animal.EmitirSom())
  }
}

func main() {
  
  cachorro := Cachorro{Animal{"Cão", "Caninus Felizis"}}
  gato := Gato{Animal{"Gato", "Gatitus Felizis"}}
  vaca := Vaca{Animal{"Vaca", "Mimosa Felizis"}}
  
  grupo := GrupoDeAnimais{}
  grupo.AdicionarAnimal(cachorro)
  grupo.AdicionarAnimal(gato)
  grupo.AdicionarAnimal(vaca)


  grupo.EmitirSons()
}

File: Questão-5/Animal.py
class Animal:
  def __init__(self, nome, especie):
    self.nome = nome
    self.especie = especie

  def EmitirSom(self):
    pass


class Cachorro(Animal):
  def EmitirSom(self):
    print("AuAu")

class Gato(Animal):
  def EmitirSom(self):
    print("Miau")

class Vaca(Animal):
  def EmitirSom(self):
    print("Muuh")

class GrupoDeAnimais(Animal):
  def __init__(self):
    self.animais = []
    
  def AdicionarAnimal(self, animal):
    self.animais.append(animal)
  
  def EmitirSomdeAnimais(self):
    for animal in self.animais:
      print(f"{animal.nome} faz:", end=' ')
      animal.EmitirSom()

    
cachorro = Cachorro("Cão", "Canis familiaris")
gato = Gato("Gatinho", "Felis catus")
vaca = Vaca("Vaca", "Mimosa Felizis")


listadeanimais = GrupoDeAnimais()

listadeanimais.AdicionarAnimal(cachorro)
listadeanimais.AdicionarAnimal(gato)
listadeanimais.AdicionarAnimal(vaca)

listadeanimais.EmitirSomdeAnimais()
